File: 06-observability/02-real-p99/cpp/coordinated_omission.cpp
// Layer 6 Topic 2 - Coordinated omission: why your load test says the p99 is fine.
//
// Why C++: it is the closest thing here to writing the load generator by hand
// against the kernel, and it shows the cost of the honest design with nothing
// hiding it. Every in-flight request in the open-loop phase is one std::thread,
// which is one pthread, which on macOS is a 512KB stack and a scheduler entry.
// The program prints how many it created. Put that number next to Go's (~0
// extra OS threads for the same hundred in flight) and Node's (none at all) and
// you have the entire practical history of why load testing tools defaulted to
// a small fixed pool of virtual users -- and therefore why an industry's worth
// of p99 numbers are measured with the generator looking away.
//
// The second thing C++ shows, by omission: the shared sample vector below is
// guarded by a std::mutex because it must be, not because anything made it so.
// Delete the lock_guard and this file still compiles, still runs, and produces
// numbers that are wrong in a way no output would reveal. The Rust version of
// this same file does not offer that option. That contrast is the point of
// having both.
//
// What this demonstrates
// ----------------------
//   * Service: single server, FIFO queue, 3ms per request -> ~333 req/s.
//   * Offered load: 200 req/s, a comfortable 60% of capacity.
//   * At T+2.5s exactly one request takes 500ms. One request.
//
//   * CLOSED-LOOP: 4 virtual users, send -> wait -> think 20ms -> repeat.
//     This is `k6 run --vus 4`, and almost every load test ever written.
//   * OPEN-LOOP: requests issued at a fixed 200/s regardless of what came back.
//     This is k6's constant-arrival-rate executor, or `vegeta -rate=200`.
//
// What to look for in the output
// ------------------------------
//   1. "requests started IN the stall window": ~4 closed-loop, ~100 open-loop.
//      That one line is the entire mechanism.
//   2. The p99 rows. Same service, same fault, two answers.
//   3. "OS threads spawned by the generator" against the Go and Node runs.
//
// Portability: standard C++17 threads only, no epoll, no /proc, nothing
// Linux-specific. Runs as written on macOS/arm64 and on Linux.
//
// Run:
//   clang++ -O2 -std=c++17 -pthread -o /tmp/coordinated_omission \
//     coordinated_omission.cpp && /tmp/coordinated_omission

#include <algorithm>
#include <atomic>
#include <chrono>
#include <condition_variable>
#include <cstdio>
#include <deque>
#include <mutex>
#include <string>
#include <thread>
#include <vector>

namespace {

using Clock = std::chrono::steady_clock;
using TimePoint = Clock::time_point;

constexpr long SERVICE_MS = 3;        // -> ~333 req/s capacity
constexpr long STALL_AFTER_MS = 2500; // when the one slow request happens
constexpr long STALL_MS = 500;        // how long that one request takes
constexpr long RUN_MS = 5000;
constexpr long OPEN_RATE_PER_SEC = 200; // offered load, ~60% of capacity
constexpr long CLOSED_VUS = 4;
constexpr long CLOSED_THINK_MS = CLOSED_VUS * 1000 / OPEN_RATE_PER_SEC;

double ms_between(TimePoint a, TimePoint b) {
    return std::chrono::duration<double, std::milli>(b - a).count();
}

/// One completed request, all offsets in milliseconds since the run started.
struct Sample {
    double sent_offset_ms;
    double latency_from_sent_ms;
    double latency_from_arrival_ms;
};

/// One unit of work. The client waits on `done`.
struct Job {
    std::mutex mu;
    std::condition_variable cv;
    bool finished = false;
    TimePoint completed_at;

    void complete(TimePoint t) {
        {
            std::lock_guard<std::mutex> guard(mu);
            completed_at = t;
            finished = true;
        }
        cv.notify_one();
    }

    TimePoint wait() {
        std::unique_lock<std::mutex> lock(mu);
        cv.wait(lock, [this] { return finished; });
        return completed_at;
    }
};

/// A single server with a FIFO queue. The queue is where the latency a
/// closed-loop generator cannot see accumulates.
class Service {
public:
    explicit Service(TimePoint epoch) : epoch_(epoch) {
        worker_ = std::thread([this] { serve(); });
    }

    void submit(Job* job) {
        {
            std::lock_guard<std::mutex> guard(mu_);
            queue_.push_back(job);
        }
        cv_.notify_one();
    }

    void stop() {
        {
            std::lock_guard<std::mutex> guard(mu_);
            running_ = false;
        }
        cv_.notify_all();
        worker_.join();
    }

private:
    void serve() {
        bool stalled = false;
        for (;;) {
            Job* job = nullptr;
            {
                std::unique_lock<std::mutex> lock(mu_);
                cv_.wait(lock, [this] { return !queue_.empty() || !running_; });
                if (queue_.empty() && !running_) return;
                job = queue_.front();
                queue_.pop_front();
            }
            double elapsed_ms = ms_between(epoch_, Clock::now());
            if (!stalled && elapsed_ms >= static_cast<double>(STALL_AFTER_MS)) {
                stalled = true;
                std::this_thread::sleep_for(std::chrono::milliseconds(STALL_MS));
            } else {
                std::this_thread::sleep_for(std::chrono::milliseconds(SERVICE_MS));
            }
            job->complete(Clock::now());
        }
    }

    TimePoint epoch_;
    std::mutex mu_;
    std::condition_variable cv_;
    std::deque<Job*> queue_;
    bool running_ = true;
    std::thread worker_;
};

struct RunResult {
    std::vector<Sample> samples;
    std::vector<double> iteration_ms;
    size_t peak_in_flight = 0;
    size_t threads_spawned = 0;
};

double percentile(std::vector<double> values, double q) {
    if (values.empty()) return 0.0;
    std::sort(values.begin(), values.end());
    size_t idx = static_cast<size_t>(q * (values.size() - 1) + 0.5);
    return values[std::min(idx, values.size() - 1)];
}

size_t started_in_stall(const std::vector<Sample>& samples) {
    size_t n = 0;
    for (const auto& s : samples) {
        if (s.sent_offset_ms >= STALL_AFTER_MS &&
            s.sent_offset_ms < STALL_AFTER_MS + STALL_MS) {
            ++n;
        }
    }
    return n;
}

std::vector<double> column(const std::vector<Sample>& samples, bool from_arrival) {
    std::vector<double> out;
    out.reserve(samples.size());
    for (const auto& s : samples) {
        out.push_back(from_arrival ? s.latency_from_arrival_ms : s.latency_from_sent_ms);
    }
    return out;
}

RunResult run_closed_loop() {
    TimePoint epoch = Clock::now();
    Service service(epoch);
    std::mutex samples_mu;   // nothing in the language requires this. It is
    std::vector<Sample> samples;  // required anyway. That is the C++ deal.
    std::vector<double> iteration_ms;

    std::vector<std::thread> users;
    for (long i = 0; i < CLOSED_VUS; ++i) {
        users.emplace_back([&] {
            while (ms_between(epoch, Clock::now()) < RUN_MS) {
                TimePoint iter_start = Clock::now();
                Job job;
                TimePoint sent = Clock::now();
                service.submit(&job);
                TimePoint done = job.wait();  // <- this virtual user is now
                                              //    blocked, and offering no load
                double latency = ms_between(sent, done);
                {
                    std::lock_guard<std::mutex> guard(samples_mu);
                    samples.push_back({ms_between(epoch, sent), latency, latency});
                }
                std::this_thread::sleep_for(std::chrono::milliseconds(CLOSED_THINK_MS));
                {
                    std::lock_guard<std::mutex> guard(samples_mu);
                    iteration_ms.push_back(ms_between(iter_start, Clock::now()));
                }
            }
        });
    }
    for (auto& t : users) t.join();
    service.stop();

    RunResult result;
    result.samples = std::move(samples);
    result.iteration_ms = std::move(iteration_ms);
    result.peak_in_flight = CLOSED_VUS;
    result.threads_spawned = CLOSED_VUS;
    return result;
}

RunResult run_open_loop() {
    TimePoint epoch = Clock::now();
    Service service(epoch);
    std::mutex samples_mu;
    std::vector<Sample> samples;
    std::atomic<size_t> in_flight{0};
    std::atomic<size_t> peak_in_flight{0};

    std::vector<std::thread> issuers;
    auto interval = std::chrono::nanoseconds(1000000000LL / OPEN_RATE_PER_SEC);
    for (long seq = 0; interval * seq < std::chrono::milliseconds(RUN_MS); ++seq) {
        TimePoint target = epoch + interval * seq;
        TimePoint now = Clock::now();
        if (target > now) std::this_thread::sleep_for(target - now);

        // One OS thread per in-flight request. This is the price of an honest
        // open-loop generator with nothing but the standard library.
        issuers.emplace_back([&, target] {
            size_t current = in_flight.fetch_add(1) + 1;
            size_t observed = peak_in_flight.load();
            while (current > observed && !peak_in_flight.compare_exchange_weak(observed, current)) {
            }
            Job job;
            TimePoint sent = Clock::now();
            service.submit(&job);
            TimePoint done = job.wait();
            {
                std::lock_guard<std::mutex> guard(samples_mu);
                samples.push_back({ms_between(epoch, sent), ms_between(sent, done),
                                   // measured from when the request was DUE
                                   ms_between(target, done)});
            }
            in_flight.fetch_sub(1);
        });
    }
    RunResult result;
    result.threads_spawned = issuers.size();
    for (auto& t : issuers) t.join();
    service.stop();

    result.samples = std::move(samples);
    result.peak_in_flight = peak_in_flight.load();
    return result;
}

}  // namespace

int main() {
    std::string bar(74, '=');
    std::printf("%s\n", bar.c_str());
    std::printf("COORDINATED OMISSION   (C++17 std::thread, single-server FIFO service)\n");
    std::printf("%s\n", bar.c_str());
    std::printf("service capacity ~%ld req/s (%ldms/request), offered load %ld req/s\n",
                1000 / SERVICE_MS, SERVICE_MS, OPEN_RATE_PER_SEC);
    std::printf("one request at T+%ldms takes %ldms instead of %ldms\n",
                STALL_AFTER_MS, STALL_MS, SERVICE_MS);
    std::printf("run length %ldms\n\n", RUN_MS);

    std::printf("running closed-loop (%ld virtual users, %ldms think time)...\n",
                CLOSED_VUS, CLOSED_THINK_MS);
    RunResult closed = run_closed_loop();
    std::printf("running open-loop (%ld req/s arrival rate)...\n\n", OPEN_RATE_PER_SEC);
    RunResult open = run_open_loop();

    std::vector<double> closed_latency = column(closed.samples, false);
    std::vector<double> open_latency = column(open.samples, true);

    std::printf("%-38s%14s%14s\n", "", "CLOSED-LOOP", "OPEN-LOOP");
    std::printf("%-38s%14zu%14zu\n", "requests completed", closed.samples.size(), open.samples.size());
    std::printf("%-38s%14zu%14zu\n", "requests started IN the stall window",
                started_in_stall(closed.samples), started_in_stall(open.samples));
    std::printf("%-38s%14zu%14zu\n", "peak requests in flight", closed.peak_in_flight, open.peak_in_flight);
    std::printf("%-38s%14zu%14zu\n", "OS threads spawned by the generator",
                closed.threads_spawned, open.threads_spawned);
    std::printf("\n");

    const char* labels[] = {"p50", "p75", "p95", "p99", "p99.9", "max"};
    const double qs[] = {0.50, 0.75, 0.95, 0.99, 0.999, 1.0};
    for (int i = 0; i < 6; ++i) {
        char label[40];
        std::snprintf(label, sizeof(label), "latency %s", labels[i]);
        std::printf("%-38s%12.1fms%12.1fms\n", label,
                    percentile(closed_latency, qs[i]), percentile(open_latency, qs[i]));
    }

    std::printf("\nThe closed-loop column measures request duration: send -> response.\n");
    std::printf("The open-loop column measures from the moment the request was DUE.\n");
    std::printf("Note the first row too: closed-loop completed %zu requests to open-loop's\n",
                closed.samples.size());
    std::printf("%zu. It did not go slower -- it asked for less, precisely while the\n",
                open.samples.size());
    std::printf("service was worst.\n");

    std::printf("\nThe tell, inside the closed-loop run alone:\n");
    std::printf("  request duration p99   : %8.1fms\n", percentile(closed_latency, 0.99));
    std::printf("  iteration duration p99 : %8.1fms\n", percentile(closed.iteration_ms, 0.99));
    std::printf("  If iteration_duration climbs while http_req_duration does not, your\n");
    std::printf("  generator stopped asking. That is k6's version of this same line.\n");

    double c99 = percentile(closed_latency, 0.99);
    double o99 = percentile(open_latency, 0.99);
    if (c99 > 0) {
        std::printf("\nVERDICT: open-loop p99 is %.1fx the closed-loop p99 for the identical\n", o99 / c99);
        std::printf("service and the identical fault.\n");
    }
    size_t hit = started_in_stall(closed.samples);
    std::printf("The closed-loop generator sampled the stall %zu times out of %zu requests\n",
                hit, closed.samples.size());
    std::printf("(%.2f%%), which is why it never reaches the 99th percentile.\n",
                100.0 * hit / std::max<size_t>(1, closed.samples.size()));
    std::printf("\nC++ footnote: the last row of the table cost one pthread and one 512KB\n");
    std::printf("stack per in-flight request. Nothing here is wrong -- it is simply the\n");
    std::printf("bill for measuring honestly without a runtime, and it is the bill that\n");
    std::printf("made closed-loop generators the industry default.\n");
    return 0;
}

File: 06-observability/02-real-p99/python/coordinated_omission.py
"""
Layer 6 Topic 2 - Coordinated omission: why your load test says the p99 is fine.

Why Python: this is the production stack, and it is also the runtime where the
open-loop generator is most expensive to build. Holding 250 requests in flight
means 250 OS threads (or 250 coroutines and an async client you may not have),
which is exactly why so many Python load tests are written closed-loop with a
small fixed worker count -- and a closed-loop test structurally cannot observe
the failure it was written to catch.

What this demonstrates
----------------------
One service. One stall. Two load generators. The same p99 question.

  * The service is a single server with a FIFO queue, 3ms per request, so its
    capacity is ~333 req/s. Offered load is 200 req/s, a comfortable 60%.
  * At T+2.5s exactly one request takes 500ms instead of 3ms. A GC pause, a
    lock, a slow downstream, a cold page. One request.

  * CLOSED-LOOP generator: 4 virtual users. Each sends a request, waits for the
    response, thinks for 30ms, repeats. This is `k6 run --vus 4`, Locust's
    default, JMeter's default, and almost every load test ever written.
  * OPEN-LOOP generator: requests are issued at a fixed 200/s regardless of
    whether earlier ones came back. This is k6's constant-arrival-rate
    executor, or `vegeta -rate=200`.

Both measure latency. They disagree, and the disagreement is not noise: during
the stall the closed-loop generator STOPS SENDING, because all six of its users
are blocked waiting. It cannot observe a queue it is not filling.

What to look for in the output
------------------------------
1. "requests started during the stall window". Closed-loop starts about 4.
   Open-loop starts about 100. That single line is the entire mechanism.
2. Closed-loop p99 versus open-loop p99. Same service, same stall.
3. Closed-loop's iteration duration versus its request duration. The stall is
   visible in the first and not the second -- that is the tell k6 gives you,
   and the reason the layer README tells you to check it.
4. Peak in-flight requests and the thread count that bought it.

Run:  python3 coordinated_omission.py
"""

import queue
import threading
import time

SERVICE_MS = 3.0        # normal service time -> ~333 req/s capacity
STALL_AFTER_MS = 2500.0  # when the one slow request happens
STALL_MS = 500.0        # how long that one request takes
RUN_MS = 5000.0
OPEN_RATE_PER_SEC = 200  # offered load, ~60% of capacity
CLOSED_VUS = 4
CLOSED_THINK_MS = CLOSED_VUS / OPEN_RATE_PER_SEC * 1000.0  # same offered load


class Request:
    __slots__ = ("seq", "arrival_ns", "sent_ns", "done_ns", "done")

    def __init__(self, seq, arrival_ns):
        self.seq = seq
        self.arrival_ns = arrival_ns   # when it *should* have been sent
        self.sent_ns = 0               # when it actually was sent
        self.done_ns = 0
        self.done = threading.Event()


class Service:
    """A single server with a FIFO queue. The queue is the point: it is where
    the latency that a closed-loop generator cannot see accumulates."""

    def __init__(self, epoch_ns):
        self.inbox = queue.Queue()
        self.epoch_ns = epoch_ns
        self.stalled = False
        self.depth_samples = []
        self._stop = object()
        self._thread = threading.Thread(target=self._serve, daemon=True)

    def start(self):
        self._thread.start()

    def stop(self):
        self.inbox.put(self._stop)
        self._thread.join()

    def submit(self, request):
        request.sent_ns = time.perf_counter_ns()
        self.inbox.put(request)

    def _serve(self):
        while True:
            item = self.inbox.get()
            if item is self._stop:
                return
            self.depth_samples.append(self.inbox.qsize())
            elapsed_ms = (time.perf_counter_ns() - self.epoch_ns) / 1e6
            if not self.stalled and elapsed_ms >= STALL_AFTER_MS:
                self.stalled = True
                time.sleep(STALL_MS / 1000.0)   # the one bad request
            else:
                time.sleep(SERVICE_MS / 1000.0)
            item.done_ns = time.perf_counter_ns()
            item.done.set()


def percentile(values, q):
    if not values:
        return 0.0
    ordered = sorted(values)
    idx = min(len(ordered) - 1, max(0, int(round(q * (len(ordered) - 1)))))
    return ordered[idx]


def in_stall_window(request, epoch_ns):
    start_ms = (request.sent_ns - epoch_ns) / 1e6
    return STALL_AFTER_MS <= start_ms < STALL_AFTER_MS + STALL_MS


def run_closed_loop():
    epoch_ns = time.perf_counter_ns()
    service = Service(epoch_ns)
    service.start()
    requests = []
    lock = threading.Lock()
    iteration_ms = []
    seq = [0]

    def virtual_user():
        while (time.perf_counter_ns() - epoch_ns) / 1e6 < RUN_MS:
            iter_start = time.perf_counter_ns()
            with lock:
                seq[0] += 1
                request = Request(seq[0], time.perf_counter_ns())
            service.submit(request)
            request.done.wait()          # <- the generator is now blocked
            with lock:
                requests.append(request)
            time.sleep(CLOSED_THINK_MS / 1000.0)
            iteration_ms.append((time.perf_counter_ns() - iter_start) / 1e6)

    users = [threading.Thread(target=virtual_user) for _ in range(CLOSED_VUS)]
    for u in users:
        u.start()
    for u in users:
        u.join()
    service.stop()

    request_ms = [(r.done_ns - r.sent_ns) / 1e6 for r in requests]
    started_in_stall = sum(1 for r in requests if in_stall_window(r, epoch_ns))
    return {
        "requests": len(requests),
        "request_ms": request_ms,
        "iteration_ms": iteration_ms,
        "started_in_stall": started_in_stall,
        "threads": CLOSED_VUS,
        "peak_in_flight": CLOSED_VUS,
    }


def run_open_loop():
    epoch_ns = time.perf_counter_ns()
    service = Service(epoch_ns)
    service.start()
    requests = []
    lock = threading.Lock()
    in_flight = [0]
    peak_in_flight = [0]
    threads = []

    def issue(request):
        with lock:
            in_flight[0] += 1
            peak_in_flight[0] = max(peak_in_flight[0], in_flight[0])
        service.submit(request)
        request.done.wait()
        with lock:
            in_flight[0] -= 1
            requests.append(request)

    interval_ns = int(1e9 / OPEN_RATE_PER_SEC)
    seq = 0
    while True:
        target_ns = epoch_ns + seq * interval_ns
        if (target_ns - epoch_ns) / 1e6 >= RUN_MS:
            break
        sleep_s = (target_ns - time.perf_counter_ns()) / 1e9
        if sleep_s > 0:
            time.sleep(sleep_s)
        seq += 1
        request = Request(seq, target_ns)
        # One thread per in-flight request. This is the cost of open-loop in a
        # thread-per-request runtime, and the reason people avoid writing it.
        t = threading.Thread(target=issue, args=(request,), daemon=True)
        threads.append(t)
        t.start()

    for t in threads:
        t.join()
    service.stop()

    # Latency measured from INTENDED arrival, not from when the generator got
    # around to sending. That difference is the whole correction.
    latency_ms = [(r.done_ns - r.arrival_ns) / 1e6 for r in requests]
    service_ms = [(r.done_ns - r.sent_ns) / 1e6 for r in requests]
    started_in_stall = sum(1 for r in requests if in_stall_window(r, epoch_ns))
    return {
        "requests": len(requests),
        "latency_ms": latency_ms,
        "service_ms": service_ms,
        "started_in_stall": started_in_stall,
        "threads": len(threads),
        "peak_in_flight": peak_in_flight[0],
    }


def main():
    print("=" * 74)
    print("COORDINATED OMISSION   (Python, single-server FIFO service)")
    print("=" * 74)
    print("service capacity ~%.0f req/s (%.0fms/request), offered load %d req/s"
          % (1000.0 / SERVICE_MS, SERVICE_MS, OPEN_RATE_PER_SEC))
    print("one request at T+%.0fms takes %.0fms instead of %.0fms"
          % (STALL_AFTER_MS, STALL_MS, SERVICE_MS))
    print("run length %.0fms\n" % RUN_MS)

    print("running closed-loop (%d virtual users, %.0fms think time)..."
          % (CLOSED_VUS, CLOSED_THINK_MS))
    closed = run_closed_loop()
    print("running open-loop (%d req/s arrival rate)...\n" % OPEN_RATE_PER_SEC)
    opened = run_open_loop()

    print("%-38s %14s %14s" % ("", "CLOSED-LOOP", "OPEN-LOOP"))
    print("%-38s %14d %14d" % ("requests completed", closed["requests"], opened["requests"]))
    print("%-38s %14d %14d" % ("requests started IN the stall window",
                               closed["started_in_stall"], opened["started_in_stall"]))
    print("%-38s %14d %14d" % ("peak requests in flight",
                               closed["peak_in_flight"], opened["peak_in_flight"]))
    print("%-38s %14d %14d" % ("OS threads used by the generator",
                               closed["threads"], opened["threads"]))
    print()
    for label, q in (("p50", 0.50), ("p75", 0.75), ("p95", 0.95),
                     ("p99", 0.99), ("p99.9", 0.999), ("max", 1.0)):
        print("%-38s %13.1fms %13.1fms"
              % ("latency " + label,
                 percentile(closed["request_ms"], q),
                 percentile(opened["latency_ms"], q)))

    print("\nThe closed-loop column measures request duration: send -> response.")
    print("The open-loop column measures from the moment the request was DUE.")
    print("That second definition is the only one a user experiences.")
    print("Note the first row too: the closed-loop run completed %d requests to"
          % closed["requests"])
    print("the open-loop run's %d. It did not go faster or slower -- it asked"
          % opened["requests"])
    print("for less, precisely while the service was worst.")

    print("\nThe tell, inside the closed-loop run alone:")
    print("  request duration p99   : %8.1fms" % percentile(closed["request_ms"], 0.99))
    print("  iteration duration p99 : %8.1fms" % percentile(closed["iteration_ms"], 0.99))
    print("  (iteration = request + think time; the stall lands here first)")
    print("  If iteration_duration climbs while http_req_duration does not, your")
    print("  generator stopped asking. That is k6's version of this same line.")

    print("\nOpen-loop, measured two ways from the same requests:")
    print("  from send    (send -> response)    p99 : %8.1fms"
          % percentile(opened["service_ms"], 0.99))
    print("  from arrival (due  -> response)    p99 : %8.1fms"
          % percentile(opened["latency_ms"], 0.99))
    print("  These agree, and that agreement is the definition of open loop: the")
    print("  generator never fell behind, so there is nothing to correct for.")
    print("  The correction only matters when the two differ -- which is exactly")
    print("  what a closed-loop generator hides, because it has no notion of a")
    print("  request being *due* at all.")

    closed99 = percentile(closed["request_ms"], 0.99)
    open99 = percentile(opened["latency_ms"], 0.99)
    if closed99 > 0:
        print("\nVERDICT: open-loop p99 is %.1fx the closed-loop p99 for the identical"
              % (open99 / closed99))
        print("service and the identical fault.")
    print("A 500ms stall in a 5s run is 10% of the wall clock. The closed-loop")
    print("generator sampled it %d times out of %d requests (%.2f%%), which is why"
          % (closed["started_in_stall"], closed["requests"],
             100.0 * closed["started_in_stall"] / max(1, closed["requests"])))
    print("it does not reach the 99th percentile at all.")


if __name__ == "__main__":
    main()
